### ISSUE TITLE
confused about generating skills in validation set

### ISSUE THREAD

**Issue by TU2021** [aiming-lab/SkillRL#8]
Hi, thanks for open-sourcing this work. I have a serious concern about the training protocol, especially the recursive skill evolution part.

From Section 3.3 and Algorithm 1, it seems that after each validation epoch, you collect failed validation trajectories and use a stronger teacher model (OpenAI o3) to generate new or refined skills. These new skills are then added to the SkillBank and used in subsequent training.

<img width="984" height="622" alt="Image" src="https://github.com/user-attachments/assets/5a7b909f-4f7b-4405-a1d1-126cd11fd11d" />

Isn’t this effectively using the validation set to update the model?

Even if you are not backpropagating through validation data, the model is still adapting to validation failures via skill updates. That looks like validation leakage. Strictly speaking, any memory or skill updates should only be done on training data. Using validation data to expand the SkillBank feels like indirectly training on validation.

Also, since you are using a much stronger model (o3) to analyze failures and generate new skills, this seems very close to iterative teacher distillation — especially when applied on validation trajectories. In practice, this is almost like having a stronger model help fix mistakes on evaluation data.

Another question: in Figure 1b, the gray curve (GRPO + Memory) — does that represent a setup without validation-based skill updates? If so, then SKILLRL is benefiting from validation-driven updates while the baseline is not, which would not be a fair comparison. Or does the gray curve also use validation failures for memory updates?

<img width="1073" height="483" alt="Image" src="https://github.com/user-attachments/assets/7cae1caf-47c5-46ab-a2dd-a497c3a56fe5" />

From a strict ML experimental perspective, all adaptive components (including teacher-based memory/skill updates) should operate only on training data. Otherwise, it becomes hard to interpret the performance gains as purely due to the method rather than protocol advantages.

Could you clarify this?

Thanks.

**Comment by richard-peng-xia** [comment]
Hi @TU2021, 

Thanks for your interest in SkillRL.

1) In our experiments, the train/validation split we used differs from the official one. Specifically, our validation set is not the same as the official validation or test set, so there is no data leakage into either of them. The writing in the paper is a bit misleading and could give the impression you described. I'll update the paper to make the data split and the skill evolution protocol clearer. I'll also push a code update to the repo to better reflect the skill updating mechanism.

2) The baseline was evaluated under the exact same data split setting. The only difference lies in the memory and skill information structures. In fact, even w/o the iterative SkillBank evolution, we can also observe a clear advantage of the skill-based architecture over the memory-based approach during training. The recursive skill evolution further amplifies this gap, but the fundamental benefit comes from the structural design itself.

**Comment by yanan1116** [comment]
i think the authors should release what these train set and validation set consist of. they should be very different queries and tasks, which means that it is task-level kind of thing. 

also, i do not know why there are so many operations on validation set and why you report success rate on validation set in figure 5. 
I never see desc about test set protocol.

i highly suspect the performance since this paper is using validation set to optimize while it is unfair for baselines.
